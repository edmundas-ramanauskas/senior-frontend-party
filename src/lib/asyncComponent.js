import React from 'react';
import PropTypes from 'prop-types';

class AsyncComponent extends React.Component {
  static propTypes = {
    load: PropTypes.func.isRequired,
  };

  state = { component: null };

  componentDidMount() {
    this.props.load().then(({ default: component }) => {
      this.setState({ component });
    });
  }

  render() {
    const { component: Component } = this.state;
    const { load, ...props } = this.props;
    return Component ? <Component {...props} /> : null;
  }
}

export default importComponent => props => (
  <AsyncComponent load={importComponent} {...props} />
);
