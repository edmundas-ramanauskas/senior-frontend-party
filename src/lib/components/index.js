export { default as Input } from './Input';
export { default as Wrapper } from './Wrapper';
export { default as Redirect } from './Redirect';
