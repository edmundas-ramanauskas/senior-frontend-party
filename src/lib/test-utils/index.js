export { default as withReduxForm } from './withReduxForm';
export { default as withReduxStore } from './withReduxStore';
export { default as withStyletron } from './withStyletron';
