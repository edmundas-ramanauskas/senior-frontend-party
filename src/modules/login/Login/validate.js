import {
  either,
  filter,
  fromPairs,
  isEmpty,
  isNil,
  keys,
  not,
  map,
  pipe,
} from 'ramda';

const required = value => ({
  valid: pipe(either(isEmpty, isNil), not)(value),
  error: 'Field is required',
});

const validators = {
  username: required,
  password: required,
};

export const fields = keys(validators);

const test = (key, value) => validators[key](value);

export default values => {
  const errors = pipe(
    map(key => [key, values[key]]),
    props => props,
    map(([key, value]) => {
      const { valid, error } = test(key, value);
      return valid ? null : [key, error];
    }),
    filter(pipe(isNil, not)),
    fromPairs,
  )(fields);
  return isEmpty(errors) ? null : errors;
};
