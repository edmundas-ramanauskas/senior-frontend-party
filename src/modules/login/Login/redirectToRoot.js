import React from 'react';
import { Redirect } from '../../../lib/components';

export default () => () => <Redirect to="/" />;
