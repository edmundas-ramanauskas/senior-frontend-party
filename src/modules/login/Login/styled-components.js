import { styled } from 'styletron-react';
import { Button as BButton } from 'react-bootstrap';
import { Input } from '../../../lib/components';

const background = require('./images/background.png');
const logo = require('./images/logo.png');

export const Wrapper = styled('div', {
  display: 'flex',
  flexDirection: 'column',
  height: '100%',
  justifyContent: 'center',
  alignsItems: 'center',
  background: `url(${background}) no-repeat center center`,
  backgroundAttachment: 'fixed',
  backgroundSize: '100% 100%',
});

export const Box = styled('div', {
  width: '361px',
  padding: '0',
  margin: '0 auto',
});

export const Logo = styled('div', {
  background: `url(${logo}) no-repeat center center`,
  height: '64px',
  marginBottom: '70px',
});

export const Field = styled(Input, {
  height: '56px !important',
  padding: '15px 16px !important',
  borderRadius: '5px !important',
  marginTop: '5px',
});

const activeStyle = {
  backgroundColor: '#86b300 !important',
};

export const Button = styled(BButton, {
  height: '56px !important',
  padding: '15px 16px !important',
  borderRadius: '5px !important',
  backgroundColor: '#9fd533',
  ':active': activeStyle,
  ':focus': activeStyle,
  ':hover': activeStyle,
  fontSize: '16px !important',
  fontWeight: '600',
});
