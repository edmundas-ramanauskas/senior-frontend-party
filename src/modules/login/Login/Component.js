import React from 'react';
import PropTypes from 'prop-types';
import { FormGroup } from 'react-bootstrap';
import { Box, Button, Field, Logo, Wrapper } from './styled-components';

const usernameIcon = require('./images/person.svg');
const passwordIcon = require('./images/lock.svg');

const Login = ({ handleSubmit }) => (
  <Wrapper>
    <Box>
      <Logo />
      <form onSubmit={handleSubmit}>
        <Field
          icon={usernameIcon}
          name="username"
          type="text"
          size="large"
          placeholder="Username"
        />
        <Field
          icon={passwordIcon}
          name="password"
          type="password"
          size="large"
          placeholder="Password"
        />
        <FormGroup>
          <Button type="submit" bsStyle="success" bsSize="large" block>
            Log In
          </Button>
        </FormGroup>
      </form>
    </Box>
  </Wrapper>
);

Login.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
};

export default Login;
