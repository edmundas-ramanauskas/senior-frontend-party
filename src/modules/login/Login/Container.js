import { isNil, pipe, prop } from 'ramda';
import { connect } from 'react-redux';
import { branch, compose } from 'recompose';
import { reduxForm, SubmissionError } from 'redux-form';
import Login from './Component';
import validate from './validate';
import redirectToRoot from './redirectToRoot';
import { FORM_ID, submit } from '../ducks';
import withToken from '../withToken';

export default compose(
  // get token from store
  withToken,
  branch(
    // if token is null
    pipe(prop('token'), isNil),
    // render login screen
    compose(
      // need to be careful with action names, because
      // reduxForm provides methods `submit` and `onSubmit`
      connect(null, {
        onFormSubmit: submit,
      }),
      reduxForm({
        form: FORM_ID,
        // set initial values for testing
        // initialValues: {
        //   username: 'tesonet',
        //   password: 'partyanimal',
        // },
        // onSubmit handler, here we do validation
        onSubmit: values =>
          new Promise(resolve => {
            const errors = validate(values);
            if (errors) {
              throw new SubmissionError(errors);
            }
            // we can pass something here to `onSubmitSuccess` callback
            resolve();
          }),
        // validation was successful
        onSubmitSuccess: (result, dispatch, { onFormSubmit }) => onFormSubmit(),
      }),
    ),
    // redirect to app root
    redirectToRoot,
  ),
)(Login);
