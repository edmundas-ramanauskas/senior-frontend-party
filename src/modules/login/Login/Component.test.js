import React from 'react';
import renderer from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { compose } from 'recompose';
import { withStyletron, withReduxForm } from '../../../lib/test-utils';
import Login from './Component';

configure({ adapter: new Adapter() });

const context = { styletron: { injectRawDeclaration: () => {} } };

describe('Login', () => {
  it('Component renders corectly', () => {
    const LoginWithProviders = compose(withStyletron, withReduxForm)(Login);
    const component = renderer.create(
      <LoginWithProviders handleSubmit={() => {}} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('All form elements are rendered', () => {
    const login = shallow(<Login handleSubmit={() => {}} />, { context });
    // find username field
    const username = login.find({ type: 'text' });
    // find password field
    const password = login.find({ type: 'password' });
    // find submit button
    const submit = login.find({ type: 'submit' });
    expect(username).toHaveLength(1);
    expect(username.prop('placeholder')).toEqual('Username');
    expect(password).toHaveLength(1);
    expect(password.prop('placeholder')).toEqual('Password');
    expect(submit).toHaveLength(1);
    expect(
      submit
        .dive()
        .find('Button')
        .dive()
        .text(),
    ).toEqual('Log In');
  });

  it('Form submit callback is triggered', done => {
    shallow(<Login handleSubmit={done} />, { context })
      .find('form')
      .simulate('submit');
  });
});
