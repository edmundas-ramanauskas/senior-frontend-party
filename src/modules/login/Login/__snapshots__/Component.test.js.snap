// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component renders corectly 1`] = `
<div
  className="undefined undefined undefined undefined undefined undefined undefined undefined"
>
  <div
    className="undefined undefined undefined"
  >
    <div
      className="undefined undefined undefined"
    />
    <form
      onSubmit={[Function]}
    >
      <div
        className="undefined form-group form-group-lg has-feedback"
      >
        <span
          className="undefined undefined undefined undefined undefined undefined undefined undefined"
        />
        <input
          className="undefined undefined undefined undefined undefined form-control"
          id={undefined}
          name="username"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          placeholder="Username"
          type="text"
          value=""
        />
      </div>
      <div
        className="undefined form-group form-group-lg has-feedback"
      >
        <span
          className="undefined undefined undefined undefined undefined undefined undefined undefined"
        />
        <input
          className="undefined undefined undefined undefined undefined form-control"
          id={undefined}
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <button
          className="undefined undefined undefined undefined undefined undefined undefined undefined undefined btn btn-lg btn-success btn-block"
          disabled={false}
          type="submit"
        >
          Log In
        </button>
      </div>
    </form>
  </div>
</div>
`;
